jerry-core/ecma/operations/ecma-promise-object.c

```diff
@@ -954,6 +954,269 @@ ecma_promise_then (ecma_value_t promise, /**< the promise which call 'then' */
   return ret;
 } /* ecma_promise_then */
 
+/**
+ * Definition of valueThunk function
+ *
+ * See also:
+ *         ES2020 25.6.5.3.1 step 8.
+ *
+ * @return ecma value
+ */
+ecma_value_t
+ecma_value_thunk_helper_cb (const ecma_value_t function_obj, /**< the function itself */
+                            const ecma_value_t this_val, /**< this_arg of the function */
+                            const ecma_value_t args_p[], /**< argument list */
+                            const ecma_length_t args_count) /**< argument number */
+{
+  JERRY_UNUSED_3 (this_val, args_p, args_count);
+
+  ecma_object_t *func_obj_p = ecma_get_object_from_value (function_obj);
+  ecma_promise_value_thunk_t *value_thunk_obj_p = (ecma_promise_value_thunk_t *) func_obj_p;
+
+  return ecma_copy_value (value_thunk_obj_p->value);
+} /* ecma_value_thunk_helper_cb */
+
+/**
+ * Definition of thrower function
+ *
+ * See also:
+ *         ES2020 25.6.5.3.2 step 8.
+ *
+ * @return ecma value
+ */
+ecma_value_t
+ecma_value_thunk_thrower_cb (const ecma_value_t function_obj, /**< the function itself */
+                             const ecma_value_t this_val, /**< this_arg of the function */
+                             const ecma_value_t args_p[], /**< argument list */
+                             const ecma_length_t args_count) /**< argument number */
+{
+  JERRY_UNUSED_3 (this_val, args_p, args_count);
+
+  ecma_object_t *func_obj_p = ecma_get_object_from_value (function_obj);
+  ecma_promise_value_thunk_t *value_thunk_obj_p = (ecma_promise_value_thunk_t *) func_obj_p;
+
+  jcontext_raise_exception (ecma_copy_value (value_thunk_obj_p->value));
+
+  return ECMA_VALUE_ERROR;
+} /* ecma_value_thunk_thrower_cb */
+
+/**
+ * Helper function for Then Finally and Catch Finally common parts
+ *
+ * See also:
+ *         ES2020 25.6.5.3.1 steps 2-7
+ *         ES2020 25.6.5.3.2 steps 2-7
+ *
+ * @return ecma value
+ */
+static ecma_value_t
+ecma_promise_than_catch_finally_helper (ecma_value_t function_obj) /**< the function itself */
```

Comment:
Pass the function pointer as argument, since that's the only difference between the catch/then finally functions.